### DISCUSSION TITLE
Running ELKS with 128K of RAM

### DISCUSSION THREAD

**Discussion by decrazyo** [jbruchon/elks#1971] [Q&A] [answered]
I've written an 8086-based emulator with 128K of RAM. The emulator can run ELKS 0.8.0-pre2 from ROM and runs `/bin/sash` without issue but I can't run any other programs. They all fail with an "Out of memory" error. Does anyone know if this is expected behavior when the system only has 128K of RAM or might there be somethings wrong with my [.config](https://github.com/user-attachments/files/16702826/config.txt)?

I've reproduced the behavior in `emu86` so I don't think it's just a bug in my emulator.

```
$ ./emu86.sh 
info: load address E0000h
info: load file elks/arch/i86/boot/Image
info: file size=10000h
success: file loaded
info: load address 80000h
info: load file image/romfs.bin
info: file size=28D3Eh
success: file loaded

ELKS Setup Headless console
PC/AT class machine, syscaps 30, 128K base ram
ELKS 0.8.0-pre2 (27424 text, 0 ftext, 3296 data, 896 bss, 61342 heap)
Kernel text e062:0, data 60:0, top 2000:0, 62K free
VFS: Mounted root device 0600 (romfs filesystem).
Can't run /bin/sh, errno -12
No init - running /bin/sh
Can't run /bin/sh, errno -12
# /bin/ls
fork failed: Out of memory 
# # sash built-in works
# ls
.
..
bin
bootopts
dev
etc
home
mnt
root
tmp
# 
```

**Comment by ghaerr** [comment, marked as answer]
Hi @decrazyo,

Pretty cool! Yes, with only 128K RAM, you've got only 62K free as indicated above in your boot message. That's not enough to run /bin/sh, although looks like it is enough for /bin/sash, but then not enough left over to fork itself to run anything else.

Notice that the kernel "heap" shows around 61k free. This can be lowered quite a bit by setting the following variable in include/linuxmt/config.h and recompiling the kernel:
```
#define SETUP_HEAPSIZE            19744         /* force kernel heap size */ // <--- set to this or a lower value, depending on resources used
```
Also, for ROM-based systems its probably a good idea to compile in the amount of RAM in use, as otherwise ELKS may try to determine it on its own using a value from setup.S, which is complicated and may not work. This can be done by setting the following in config.h:
```
#define SETUP_MEM_KBYTES        256     /* force available memory in 1K bytes */ //<-- set to your RAM size
```

(There are different sections within config.h, make sure you set the above within the CONFIG_ROMCODE section).

There are quite a few other ways to tune the system once you have this running, see include/linuxmt/limits.h.

It looks like you're using `emu86`. There's a way to set that emulator's emulated memory as well, if needed.

Thank you!
